From c4b447b9fead4b2d655c9bc2ef78d4ef64035f8c Mon Sep 17 00:00:00 2001
From: Jeffy Chen <jeffy.chen@rock-chips.com>
Date: Wed, 1 Jun 2022 11:28:24 +0800
Subject: [PATCH 71/99] desktop-shell: Delay locking when having pending
 fade-out animations

Avoid "unexpectedly large timestamp jump" warning when resuming with
multi-head.

Signed-off-by: Jeffy Chen <jeffy.chen@rock-chips.com>
---
 desktop-shell/shell.c         | 13 ++++++++++++-
 include/libweston/libweston.h |  2 ++
 2 files changed, 14 insertions(+), 1 deletion(-)

diff --git a/desktop-shell/shell.c b/desktop-shell/shell.c
index 41b4fcb..645b765 100644
--- a/desktop-shell/shell.c
+++ b/desktop-shell/shell.c
@@ -3857,6 +3857,10 @@ static void
 shell_fade_done(struct weston_view_animation *animation, void *data)
 {
 	struct desktop_shell *shell = data;
+	struct weston_compositor *compositor = shell->compositor;
+
+	if (shell->fade.type == FADE_OUT)
+		compositor->pending_fade_out --;
 
 	if (!shell->fade.curtain)
 		return;
@@ -3868,7 +3872,8 @@ shell_fade_done(struct weston_view_animation *animation, void *data)
 		shell->fade.curtain = NULL;
 		break;
 	case FADE_OUT:
-		lock(shell);
+		if (!compositor->pending_fade_out)
+			lock(shell);
 		break;
 	default:
 		break;
@@ -3961,6 +3966,9 @@ shell_fade(struct desktop_shell *shell, enum fade_type type)
 	if (shell->fade.animation) {
 		weston_fade_update(shell->fade.animation, tint);
 	} else {
+		if (type == FADE_OUT)
+			shell->compositor->pending_fade_out ++;
+
 		shell->fade.animation =
 			weston_fade_run(shell->fade.curtain->view,
 					1.0 - tint, tint, 300.0,
@@ -4860,6 +4868,9 @@ shell_destroy(struct wl_listener *listener, void *data)
 	if (shell->fade.animation) {
 		weston_view_animation_destroy(shell->fade.animation);
 		shell->fade.animation = NULL;
+
+		if (shell->fade.type == FADE_OUT)
+			shell->compositor->pending_fade_out --;
 	}
 
 	if (shell->fade.curtain) {
diff --git a/include/libweston/libweston.h b/include/libweston/libweston.h
index 1886cef..782464c 100644
--- a/include/libweston/libweston.h
+++ b/include/libweston/libweston.h
@@ -1542,6 +1542,8 @@ struct weston_compositor {
 	bool warm_up;
 
 	bool hide_cursor;
+
+	uint32_t pending_fade_out;
 };
 
 struct weston_solid_buffer_values {
-- 
2.20.1

